---
title: "ST_pipeline"
output: html_document
---
Data: https://support.10xgenomics.com/spatial-gene-expression/datasets/1.0.0/V1_Breast_Cancer_Block_A_Section_1?
File contents help: https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/output/images
NN tutorial: https://blog.rstudio.com/2017/09/05/keras-for-r/


You MUST first unpack the gz files!!
```{r setup, include=FALSE}
#devtools::install_github("satijalab/seurat", ref = "spatial")
library(backports)
library(Seurat)
#library(SeuratData)
library(ggplot2)
library(patchwork)
library(dplyr)


library(Matrix)
library(rjson)
library(cowplot)
library(RColorBrewer)
library(grid)
library(readbitmap)
library(hdf5r)
library(data.table)


library(raster)


library(keras)
library(EBImage)
library(stringr)
library(pbapply)
```

Loading in spot position csv
```{r}
# Load in spot location CSV file(doesn't have headers)
image_location <- read.csv("data/spatial/tissue_positions_list.csv", header=FALSE)
names(image_location) <- c("barcode", "in_tissue", "array_row", "array_col", "pxl_col_in_fullres_yValue", "pxl_row_in_fullres_xValue")
```
 
Loading in the Whole Slide Image (WSI)
```{r}
file_name <- 'data/V1_Breast_Cancer_Block_A_Section_1_image.tif'

#creates a raster object from the tiff image file
#imported_raster <- raster(str_name)
wsi <- brick(file_name)

#check to see the image contains the correct amount of pixels
dim(wsi)
plotRGB(wsi)

e <- extent(13091, 13341, 4639, 4889)
cropped_img <- crop(wsi, e)
plotRGB(cropped_img)
```


Defining a function to crop the spot images according to their centers
```{r}
# grabbing only the spots with tissue
image_location <- image_location %>%
  filter(in_tissue == 1)

#defining a function to get a cropped spot image given the x and y coordinates of the center of the spot
crop_spot <- function(img, x, y){
e <- extent(x, #using 136 since it is ~half the average distance (273 pixels) between spot centers. We could consider using a different measure to more closely fit to the spot ratius.
              x + 1,
              y,
              y + 1)
cropped_img <- crop(img, e)
return(cropped_img)
}
```


Partitioning the spots into testing and training data using a random sampling
(try implementing combine from here: https://www.youtube.com/watch?v=LxTDLEdetmI)
```{r}
set.seed(1998)
#nrow should be used when looking at the entire data set and no longer restricting to 500 spots
index <- sample(2,
                nrow(image_location),
                replace = TRUE,
                prob = c(0.7, 0.3)) #splits data into testing and training

training_data <- array(NA, dim=c(sum(index == 1), 1, 1, 3))
index_count <- 1 
training_indexes <- which(index == 1)
print("Collecting training data")
pb <- txtProgressBar(min = 0, max = length(training_indexes), style = 3)
for (i in training_indexes) #This for loop takes about an hour and a half when run on all spots
{
  cropped_img <- crop_spot(wsi, 
                           image_location$pxl_row_in_fullres_xValue[i], 
                           image_location$pxl_col_in_fullres_yValue[i])
  #plotRGB(cropped_img) 
  cropped_img <- as.array(cropped_img)
  training_data[index_count,,,] <- cropped_img
  setTxtProgressBar(pb, index_count)
  index_count <- index_count + 1
}
close(pb)
#dim(training_data)


testing_data <- array(NA, dim=c(sum(index == 2), 1, 1, 3))
index_count <- 1
testing_indexes <- which(index == 2)
print("Collecting testing data")
pb <- txtProgressBar(min = 0, max = length(testing_indexes), style = 3)
for (i in testing_indexes) #This for loop takes about 40 mins when run on all spots
{
  cropped_img <- crop_spot(wsi, 
                           image_location$pxl_row_in_fullres_xValue[i], 
                           image_location$pxl_col_in_fullres_yValue[i])
  #plotRGB(cropped_img)
  cropped_img <- as.array(cropped_img)
  testing_data[index_count,,,] <- cropped_img
  setTxtProgressBar(pb, index_count)
  index_count <- index_count + 1
}
close(pb)
#dim(testing_data)
```


Loading in gene expression data and formatting the indexes to their corresponding spot image
```{r}
feature_matrix <- Load10X_Spatial(data.dir = "/Users/colten/Desktop/Perkins_Lab_ST/Identifying (usable) image spots/data",
                         filename = "V1_Breast_Cancer_Block_A_Section_1_filtered_feature_bc_matrix.h5",
                         assay = "Spatial",
                         slice = "slice1",
                         filter.matrix = TRUE,
                         to.upper = FALSE)

feature_matrix <- SCTransform(feature_matrix, assay = "Spatial", verbose = FALSE)

SCT_barcode_feature_matrix <- t(GetAssayData(object = feature_matrix, slot = "scale.data"))

barcodes <- rownames(SCT_barcode_feature_matrix)
features <- colnames(SCT_barcode_feature_matrix)

indexes_of_training_barcodes <- match(image_location$barcode[training_indexes], barcodes)
training_barcode_feature_matrix <- t(SCT_barcode_feature_matrix[indexes_of_training_barcodes,])
dim(training_barcode_feature_matrix)


indexes_of_testing_barcodes <- match(image_location$barcode[testing_indexes], barcodes)
testing_barcode_feature_matrix <- t(SCT_barcode_feature_matrix[indexes_of_testing_barcodes,])
dim(testing_barcode_feature_matrix)

indexes_of_testing_barcodes <- match(image_location$barcode[testing_indexes], barcodes)
testing_barcode_feature_matrix <- t(barcode_feature_matrix[,indexes_of_testing_barcodes])
dim(testing_barcode_feature_matrix)
```






Linear regression testing
```{r}
d1 <- c()
d2 <- c()
d3 <- c()

for (i in 1:dim(training_data)[1]) {
  d1 <- append(d1, training_data[i,,,1]) 
  d2 <- append(d2, training_data[i,,,2]) 
  d3 <- append(d3, training_data[i,,,3]) 
}

training_barcode_feature_matrix <- training_barcode_feature_matrix %>%
  as.matrix() %>%
  cbind(red = d1,
         green = d2,
         blue = d3) %>%
  as.data.frame()

# Multiple Linear Regression Example
fit <- lm(ISG15 ~ red + green + blue, data=training_barcode_feature_matrix)
summary(fit) # show results
```

